v1/tarball/layer.go

```diff
@@ -64,33 +64,40 @@ func (l *layer) Size() (int64, error) {
 
 // LayerFromFile returns a v1.Layer given a tarball
 func LayerFromFile(path string) (v1.Layer, error) {
-	file, err := os.Open(path)
+	opener := func() (io.ReadCloser, error) {
+		return os.Open(path)
+	}
+	return LayerFromOpener(opener)
+}
+
+// LayerFromFile returns a v1.Layer given an Opener function
+func LayerFromOpener(opener Opener) (v1.Layer, error) {
```

Comment:
There might be a better existing interface here we could take, but I'm not sure of one. Essentially we need a reader that can be reset to the beginning repeatedly. An opened file works, because we can Seek it back to 0. A slice of bytes works because we can just create a new reader at 0 each time we need to reset.

I'm not sure of an interface (other than this one where take a function to perform the "reset") that can cover both of these.

Reply:
https://golang.org/pkg/io/#ReadSeeker ?

Reply:
And bytes.Reader implements it: https://golang.org/pkg/bytes/#Reader.Seek

Reply:
Actually looking at this again we need more than a Reset/Seek(0,0), because of the Compressed/Uncompressed methods on Layer. A single file handle won't be safe during concurrent access.